core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -269,19 +269,22 @@ class ReplicaFetcherThread(name: String,
    * Truncate the log for each partition's epoch based on leader's returned epoch and offset.
    * The logic for finding the truncation offset is implemented in AbstractFetcherThread.getOffsetTruncationState
    */
-  override def truncate(tp: TopicPartition, offsetTruncationState: OffsetTruncationState): Unit = {
-    val replica = replicaMgr.localReplicaOrException(tp)
-    val partition = replicaMgr.getPartition(tp).get
-    partition.truncateTo(offsetTruncationState.offset, isFuture = false)
-
-    if (offsetTruncationState.offset < replica.highWatermark.messageOffset)
-      warn(s"Truncating $tp to offset ${offsetTruncationState.offset} below high watermark " +
-        s"${replica.highWatermark.messageOffset}")
-
-    // mark the future replica for truncation only when we do last truncation
-    if (offsetTruncationState.truncationCompleted)
-      replicaMgr.replicaAlterLogDirsManager.markPartitionsForTruncation(brokerConfig.brokerId, tp,
-        offsetTruncationState.offset)
+  override def truncate(tp: TopicPartition, offsetTruncationState: OffsetTruncationState,
+                        offsetTruncationStates: Option[Map[TopicPartition, OffsetTruncationState]] = None): Unit = {
+    offsetTruncationStates.getOrElse(Map(tp -> offsetTruncationState)).foreach { case (tp, offsetTruncationState) =>
+      val replica = replicaMgr.localReplicaOrException(tp)
+      val partition = replicaMgr.getPartition(tp).get
+      partition.truncateTo(offsetTruncationState.offset, isFuture = false)
```

Comment:
This is trickier than I thought. Since partition.truncateTo() holds a leaderIsrUpdateLock, it's a bit hard to truncate a batch of partitions together. Perhaps what we can do is to call LogManager.truncateTo() on each partition separately, but w/o calling checkpointLogRecoveryOffsets there. After all partitions are truncated, we call LogManager.checkpointLogRecoveryOffsets() on a batch of affected logs.

Reply:
hmmm.... I could do that. However, if calling `checkpointLogRecoveryOffsets` after all partitions are truncated is the way to go,  we still have to change the `truncate` signature of `AbstractFetcherThread` to have it support batch truncating since AbstractFetcherThread does not see `logManager` or `replicaManager` therefore it cannot invoke `checkpointLogRecoveryOffsets` directly. I do think it is not a good idea to do so...

Reply:
Yes, in order to batch the checkpoint of recovery offsets, we will need to change the signature of AbstractFetcherThread.truncate(). This probably requires a bit more thought and is beyond what this jira was originally intended for. So, if you prefer, we can just revert the changes in AbstractFetcherThread/ReplicaFetcherThread/ReplicaAlterLogDirsThread and commit the rest of the fixes in the PR. I will file a separate jira to address the issue of batching the checkpoint of recovery offsets.